compiler/src/iree/compiler/Dialect/HAL/Analysis/DeviceAnalysis.cpp

```diff
@@ -0,0 +1,234 @@
+      Position::forValue(deviceValue));
+  if (!globalPVS || !globalPVS->isValidState() ||
+      globalPVS->isUndefContained()) {
+    return std::nullopt;
+  }
+  SetVector<IREE::Util::GlobalOpInterface> globalOps;
+  for (auto globalOp : globalPVS->getAssumedSet()) {
+    globalOps.insert(globalOp);
+  }
+  return globalOps;
+}
+
+std::optional<DeviceSet>
+DeviceAnalysis::lookupDeviceTargets(Value deviceValue) {
+  auto valuePVS = solver.lookupElementFor<DeviceTargetValuePVS>(
+      Position::forValue(deviceValue));
+  if (!valuePVS || !valuePVS->isValidState() || valuePVS->isUndefContained()) {
+    return defaultDeviceSet;
+  }
+  return DeviceSet(valuePVS->getAssumedSet());
+}
+
+// Returns a set of target devices that may be active for the given
+// operation. This will recursively walk parent operations until one with
+// the `hal.device.targets` attribute is found.
+//
+// This is a legacy mechanism for performing the search. Newer code should use
+// affinities or !hal.device analysis instead.
+static void gatherLegacyDeviceTargetAttrs(
+    Operation *op, SetVector<IREE::HAL::DeviceTargetAttr> &resultSet) {
+  auto attrId = StringAttr::get(op->getContext(), "hal.device.targets");
+  while (op) {
+    auto targetsAttr = op->getAttrOfType<ArrayAttr>(attrId);
+    if (targetsAttr) {
+      for (auto elementAttr : targetsAttr) {
+        if (auto targetAttr =
+                dyn_cast<IREE::HAL::DeviceTargetAttr>(elementAttr)) {
+          resultSet.insert(targetAttr);
+        } else {
+          // HACK: this legacy approach is deprecated and only preserved for
+          // existing behavior. It's ok to get angry here as users should not be
+          // trying to use this pass prior to device materialization.
+          assert(false &&
+                 "legacy hal.device.targets only support hal.device.targets");
+        }
+      }
+      return;
+    }
+    op = op->getParentOp();
+  }
+  // No devices found; let caller decide what to do.
+}
+
+// Recursively resolves the referenced device into targets.
+void DeviceAnalysis::gatherDeviceTargets(
+    Attribute rootAttr, Operation *fromOp,
+    SetVector<IREE::HAL::DeviceTargetAttr> &resultSet) {
+  SetVector<Attribute> worklist;
+  worklist.insert(rootAttr);
+  do {
+    auto attr = worklist.pop_back_val();
+    if (!TypeSwitch<Attribute, bool>(attr)
+             .Case<SymbolRefAttr>([&](auto symRefAttr) {
+               auto globalOp =
+                   explorer.getSymbolTables()
+                       .lookupNearestSymbolFrom<IREE::Util::GlobalOpInterface>(
+                           fromOp, symRefAttr);
+               assert(globalOp && "global reference must be valid");
+               if (auto initialValueAttr = globalOp.getGlobalInitialValue()) {
+                 // Global with a device initialization value we can analyze.
+                 worklist.insert(initialValueAttr);
+                 return true;
```

Comment:
clang-format made quite the staircase with indentation here ._.

Reply:
yeah, it's pretty hilarious - I try to avoid it when I can but when there are a lot of captures it gets a lot less readable to have functions that take a dozen args ;(